File: go/internal/adapters/payments.go
package adapters

import (
	"context"
	"fmt"
	"sync"

	"warehouse/internal/application"
	"warehouse/internal/domain"
)

type chargeOutcome struct {
	receipt application.ChargeReceipt
	err     error
}

// FakePaymentProcessor is an application.PaymentProcessor double that records
// every charge attempt. Configure decline to make each collection fail with
// PaymentDeclined. Identical idempotency keys replay the original outcome.
type FakePaymentProcessor struct {
	mu            sync.Mutex
	decline       bool
	chargedOrders []*domain.Order
	receipts      map[string]chargeOutcome
	refunded      []application.ChargeReceipt
	FailRefund    bool
}

// NewFakePaymentProcessor starts in the configured outcome mode with an empty
// attempt log.
func NewFakePaymentProcessor(decline bool) *FakePaymentProcessor {
	return &FakePaymentProcessor{
		decline:  decline,
		receipts: make(map[string]chargeOutcome),
	}
}

// Charges returns the orders passed so far, in attempt order.
func (p *FakePaymentProcessor) Charges() []*domain.Order {
	p.mu.Lock()
	defer p.mu.Unlock()
	out := make([]*domain.Order, len(p.chargedOrders))
	copy(out, p.chargedOrders)
	return out
}

// Refunds returns receipts voided so far, in attempt order.
func (p *FakePaymentProcessor) Refunds() []application.ChargeReceipt {
	p.mu.Lock()
	defer p.mu.Unlock()
	out := make([]application.ChargeReceipt, len(p.refunded))
	copy(out, p.refunded)
	return out
}

// Charge records the attempt, then honors the configured outcome. A retry
// with the same key returns the original receipt or decline.
func (p *FakePaymentProcessor) Charge(_ context.Context, order *domain.Order, idempotencyKey string) (application.ChargeReceipt, error) {
	p.mu.Lock()
	defer p.mu.Unlock()
	if existing, ok := p.receipts[idempotencyKey]; ok {
		if existing.err != nil {
			return existing.receipt, fmt.Errorf("charge: %w", existing.err)
		}
		return existing.receipt, nil
	}
	p.chargedOrders = append(p.chargedOrders, order)
	if p.decline {
		declined := domain.PaymentDeclined{Reason: fmt.Sprintf("payment declined for order %s", order.ID().Value)}
		p.receipts[idempotencyKey] = chargeOutcome{err: declined}
		return application.ChargeReceipt{}, fmt.Errorf("charge: %w", declined)
	}
	receipt := application.ChargeReceipt{OrderID: order.ID(), IdempotencyKey: idempotencyKey}
	p.receipts[idempotencyKey] = chargeOutcome{receipt: receipt}
	return receipt, nil
}

// Refund voids a prior successful charge.
func (p *FakePaymentProcessor) Refund(_ context.Context, receipt application.ChargeReceipt) error {
	p.mu.Lock()
	defer p.mu.Unlock()
	if p.FailRefund {
		return fmt.Errorf(
			"refund: %w",
			domain.CompensationFailure{Stage: "refund", Detail: "forced failure"},
		)
	}
	p.refunded = append(p.refunded, receipt)
	return nil
}

// compile-time proof that the adapter satisfies its port.
var _ application.PaymentProcessor = (*FakePaymentProcessor)(nil)

File: swift/Sources/Warehouse/Adapters/Payments.swift
import Dispatch

/// PaymentProcessor test double that records every charge attempt.
public final class FakePaymentProcessor: PaymentProcessor, @unchecked Sendable {
  private let queue = DispatchQueue(label: "warehouse.payments")
  private let decline: Bool
  private var receipts: [String: Result<ChargeReceipt, PaymentDeclined>] = [:]
  public private(set) var chargedOrders: [Order] = []
  public private(set) var refunded: [ChargeReceipt] = []
  public var failRefund = false

  public init(decline: Bool = false) {
    self.decline = decline
  }

  public func charge(
    _ order: Order,
    idempotencyKey: String
  ) -> Result<ChargeReceipt, PaymentDeclined> {
    queue.sync {
      if let existing = receipts[idempotencyKey] {
        return existing
      }
      chargedOrders.append(order)
      if decline {
        let declined = PaymentDeclined("payment declined for order \(order.id.value)")
        receipts[idempotencyKey] = .failure(declined)
        return .failure(declined)
      }
      let receipt = ChargeReceipt(orderId: order.id, idempotencyKey: idempotencyKey)
      receipts[idempotencyKey] = .success(receipt)
      return .success(receipt)
    }
  }

  public func refund(_ receipt: ChargeReceipt) -> Result<Void, CompensationFailure> {
    queue.sync {
      if failRefund {
        return .failure(CompensationFailure(stage: "refund", detail: "forced failure"))
      }
      refunded.append(receipt)
      return .success(())
    }
  }
}
